### DISCUSSION TITLE
[v4] Custom utilities result in duplicate CSS output

### DISCUSSION THREAD

**Discussion by kieranm** [tailwindlabs/tailwindcss#14923] [Help] [answered]
Hey all,

Noticed some strange behaviour in V4 and wanted to check if it was a bug or not.

If you define a custom utility like this:
**shades.css**
```
@utility shades {
  --shade-0: 100% calc(0 * var(--chroma-multiplier, 1));
  --shade-1: 99.27% calc(0.003 * var(--chroma-multiplier, 1));
  --shade-2: 98.5% calc(0.007 * var(--chroma-multiplier, 1));
  --shade-3: 96.94% calc(0.015 * var(--chroma-multiplier, 1));
  --shade-4: 93.89% calc(0.031 * var(--chroma-multiplier, 1));
  --shade-5: 88.22% calc(0.062 * var(--chroma-multiplier, 1));
  --shade-6: 78.71% calc(0.116 * var(--chroma-multiplier, 1));
  --shade-7: 65.18% calc(0.148 * var(--chroma-multiplier, 1));
  --shade-8: 53.43% calc(0.121 * var(--chroma-multiplier, 1));
  --shade-9: 45.57% calc(0.103 * var(--chroma-multiplier, 1));
  --shade-10: 38.51% calc(0.087 * var(--chroma-multiplier, 1));
  --shade-11: 32.55% calc(0.074 * var(--chroma-multiplier, 1));
  --shade-12: 27.79% calc(0.063 * var(--chroma-multiplier, 1));
  --shade-13: 24.15% calc(0.055 * var(--chroma-multiplier, 1));
  --shade-14: 21.45% calc(0.048 * var(--chroma-multiplier, 1));
  --shade-15: 0% calc(0 * var(--chroma-multiplier, 1));
}

```

Then apply it to the body as follows:
**main.css**
```
@import "shades.css"
body {
  @apply shades
}
```

Now assume you import `shades.css` in a couple of other places too.

For each import you add, the styles defined in this utility seem to be duplicated even though they have only been `@apply`ed to `body` once. If you have 5 `@import "shades.css"` statements dotted around your files the @apply seems to happen 5 times and you see duplicate warnings in the console.

<img width="481" alt="Screenshot 2024-11-08 at 15 58 03" src="https://github.com/user-attachments/assets/a65b66c5-fedf-40ed-a92f-bf2bcb3547ae">

Is this a bug?

Thanks!

**Comment by philipp-spiess** [comment, marked as answer]
Hey! This is actually on purpose, if you add a utility via `@utility`, it's going to be added in addition to the existing utilities and does not overwrite it. This is done to improve backwards compatibility with v3 which had similar behavior.

Check out this example without using `@apply`:

<img width="1261" alt="Screenshot 2024-11-08 at 17 22 29" src="https://github.com/user-attachments/assets/5469c6ec-be56-484f-9529-bb4c6d050b59">

https://play.tailwindcss.com/nZzU6dUCpb
